### PR TITLE
Bugfix BTS-1243

### PR DESCRIPTION
**Pull request by MBkkt**
### Scope & Purpose

See comments

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Metrics/ClusterMetricsFeature.cpp [hunk 1]
```diff
@@ -262,11 +277,15 @@ bool ClusterMetricsFeature::readData(futures::Try<LeaderResponse>&& raw) {
     return false;
   }
   velocypack::Slice metrics{raw.get()->data()};
+  if (metrics.isNull()) {
+    return true;  // our data is up to date
+  }
```
Comment (MBkkt) [inline arangod/Metrics/ClusterMetricsFeature.cpp:282]: It's main part of fix.
If we receive null (none in old code, null used now because it more unique than zero byte)
it means our data is up to date and we don't need to try to repeat fetch data from leader

File: arangod/Metrics/ClusterMetricsFeature.cpp [hunk 2]
```diff
@@ -239,9 +252,11 @@ bool ClusterMetricsFeature::writeData(uint64_t version,
     return false;
   }
   auto metrics = parse(std::move(raw).get());
-  if (metrics.values.empty()) {
+  bool const currEmpty = metrics.values.empty();
+  if (currEmpty && _prevEmpty) {
     return true;
   }
```
Comment (MBkkt) [inline arangod/Metrics/ClusterMetricsFeature.cpp:258]: It's other fix.
If for some reason we remove all data, we need to remove it from our leader state.
I can just remove this if, but I think if data is constantly empty it's not a good idea to reassemble it and move version to the future.

File: arangod/Metrics/ClusterMetricsFeature.cpp [hunk 3]
```diff
@@ -336,18 +355,24 @@ void ClusterMetricsFeature::toPrometheus(std::string& result,
...
   for (size_t i = 0; i < size; i += 3) {
-    auto name = metrics.at(i).stringView();
-    auto labels = metrics.at(i + 1).stringView();
-    auto value = metrics.at(i + 2).getNumber<uint64_t>();
+    auto name = (*metrics).stringView();
+    metrics.next();
+    auto labels = (*metrics).stringView();
+    metrics.next();
+    auto value = (*metrics).getNumber<uint64_t>();
+    metrics.next();
```
Comment (MBkkt) [inline arangod/Metrics/ClusterMetricsFeature.cpp:375]: Just a minor parsing speedup

File: arangod/RestHandler/RestMetricsHandler.cpp [hunk 4]
```diff
@@ -211,7 +209,7 @@ RestStatus RestMetricsHandler::execute() {
     if (isOutdated(*_request, data)) {
       _response->addPayload(VPackSlice{data->packed->data()});
     } else {
-      _response->addPayload(VPackSlice::noneSlice());
+      _response->addPayload(VPackSlice::nullSlice());
```
Comment (MBkkt) [inline arangod/RestHandler/RestMetricsHandler.cpp:212]: null used instead of none, because it's more unique value

File: arangod/Metrics/ClusterMetricsFeature.cpp [hunk 5]
```diff
@@ -199,18 +213,17 @@ void ClusterMetricsFeature::update() {
...
+  auto rebootId = oldData.get("RebootId").getNumber<uint64_t>();
+  auto serverId = oldData.get("ServerId").copyString();
   data.reset();
   metricsFromLeader(nf, cf, *leader, std::move(serverId), rebootId, version)
       .thenFinal([this](futures::Try<LeaderResponse>&& raw) mutable noexcept {
         if (wasStop()) {
           return;
         }
-        uint32_t timeoutMs = 1;  // invalid leader, immediately retry
+        uint32_t timeoutMs = 100;  // invalid leader, immediately retry
```
Comment (MBkkt) [inline arangod/Metrics/ClusterMetricsFeature.cpp:226]: I set 100 instead of 1.
Because I think repeat request every 0.1 seconds is often enough.
For cases when something goes wrong in our setup, for an example leader reelection because network partion/leader crash.

File: arangod/Metrics/ClusterMetricsFeature.cpp [hunk 6]
```diff
@@ -199,18 +213,17 @@ void ClusterMetricsFeature::update() {
         });
   }
   if (leader->empty()) {
-    return repeatUpdate(1);  // invalid leader, immediately retry
+    return repeatUpdate(100);  // invalid leader, immediately retry
```
Comment (Dronplane) [inline arangod/Metrics/ClusterMetricsFeature.cpp:216]: Suggest to use here the configured timeout for metrics

File: arangod/Metrics/ClusterMetricsFeature.cpp (hunk 5 above)
Comment (Dronplane) [reply arangod/Metrics/ClusterMetricsFeature.cpp:226]: Suggest to use here the configured timeout for metrics

**Review by Dronplane — CHANGES_REQUESTED** [review]
Please add a least tests with Empty data being accepted and not re-requested if that is not much effort.

**Review by Dronplane — APPROVED** [review]
LGTM.

**Comment by Dronplane** [comment]
Tests blue except unrelated failures with tsan
